src/System.Net.Security/src/System/Net/Security/SniHelper.cs

```diff
@@ -0,0 +1,337 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Globalization;
+using System.Text;
+
+namespace System.Net.Security
+{
+    internal class SniHelper
+    {
+        private const int ProtocolVersionSize = 2;
+        private const int UInt24Size = 3;
+        private const int RandomSize = 32;
+        private static IdnMapping s_idnMapping = CreateIdnMapping();
+
+        public static string GetServerName(byte[] clientHello)
+        {
+            return GetSniFromSslPlainText(clientHello);
+        }
+
+        private static string GetSniFromSslPlainText(ReadOnlySpan<byte> sslPlainText)
+        {
+            // https://tools.ietf.org/html/rfc6101#section-5.2.1
+            const int ContentTypeOffset = 0;
+            const int ProtocolVersionOffset = ContentTypeOffset + sizeof(ContentType);
+            const int LengthOffset = ProtocolVersionOffset + ProtocolVersionSize;
+            const int HandshakeOffset = LengthOffset + sizeof(ushort);
+
+            // SSL v2's ContentType has 0x80 bit set.
+            // We do not care about SSL v2 here because it does not support client hello extensions
+            if (sslPlainText.Length < HandshakeOffset || (ContentType)sslPlainText[ContentTypeOffset] != ContentType.Handshake)
+            {
+                return null;
+            }
+
+            // Skip ContentType and ProtocolVersion
+            int handshakeLength = BinaryPrimitives.ReadUInt16BigEndian(sslPlainText.Slice(LengthOffset));
+            ReadOnlySpan<byte> sslHandshake = sslPlainText.Slice(HandshakeOffset);
+
+            if (handshakeLength != sslHandshake.Length)
+            {
+                return null;
+            }
+
+            return GetSniFromSslHandshake(sslHandshake);
+        }
+
+        private static string GetSniFromSslHandshake(ReadOnlySpan<byte> sslHandshake)
+        {
+            // https://tools.ietf.org/html/rfc6101#section-5.6
```

Comment:
I didn't mean that you should delete the structure documentation, I meant that you should put it next to the code that reads it instead of looking like API docs.

Reply:
@bartonjs I didn't find it useful anymore considering all the used offsets are in the code now and they are calculated with sizeof/constant which contains the name of the type

Reply:
That makes your code more legible, yes.  But it's almost always good to have a close reference to what is being read in parsing routines.  Once you've written the parser you probably don't need the reference, but reviewers and future maintainers probably want it.